README
NAME
    MooseX::Deprecated - mark attributes and methods as deprecated

SYNOPSIS
       package Goose
       {
          use Moose;
      
          has feathers => (is => 'ro');
      
          sub honk { say "Honk!" }
      
          with "MooseX::Deprecated" => {
             attributes => [ "feathers" ],
             methods    => [ "honk" ],
          };
       }

DESCRIPTION
    MooseX::Deprecated is a parameterizable role that makes it easy to
    deprecate particular attributes and methods in a class.

    In the SYNOPSIS above, `before` method modifiers will be installed on the
    `feathers` accessor and the `honk` method, issuing a deprecation warning.
    Additionally, an `after` modifier will be installed on the class' `BUILD`
    method which will issue deprecation warnings for any deprecated attributes
    passed to the constructor.

    The warning text will be something along the lines of: "%s is a deprecated
    %s"

    Warnings are issued in the "deprecated" warnings category, so can be
    disabled using:

       no warnings qw( deprecated );

    Warnings can be upgraded to fatal errors with:

       use warnings FATAL => qw( deprecated );

    When consuming the role you *must* pass either a list of attributes, or a
    list of methods, or both, as parameters to the role. If you forget to do
    so, you'll get an error message like: "%s with no list of attributes or
    methods".

CAVEATS
    To deprecate an attribute, the attribute must actually exist at the time
    you consume this role. In particular, this will not work:

       package Goose
       {
          use Moose;
      
          with "MooseX::Deprecated" => {
             attributes => [ "feathers" ],
          };
      
          has feathers => (is => 'ro');
       }

    Because the "feathers" attribute isn't defined until *after* the role is
    consumed. Attempting the above will die with a nasty error message:
    "Attribute %s does not exist in %s so cannot be deprecated".

    If a deprecated attribute handles any methods via delegation, then calling
    these methods will result in not one, but two warnings. One warning for
    calling the delegated method; the other warning for calling the accessor
    (reader) to obtain the object to delegate to. This could theoretically be
    changed, but I'm comfortable with the existing situation.

    Warnings issued by the accessor (reader) during method delegation come
    from inside your class, and thus the caller cannot disable them or
    fatalize them.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=MooseX-Deprecated>.

  Perl 5.8
    The behaviour of `warnings::warnif` changed significantly between Perl 5.8
    and 5.10; the location considered to be the effective caller changed from
    being "like `warn`" to being "like `carp`" (the latter being considerably
    more useful). Therefore under Perl 5.8, doing things like `no warnings
    "deprecated"` in your code to control warnings from this role is rather
    useless, because your code is unlikely to be considered to be the caller.

    In the test suite I just skip the complex test that checks for this on
    Perl prior to 5.10, allowing you to install this module without a hitch on
    Perl 5.8.

SEE ALSO
    Package::DeprecationManager provides a more powerful and complicated set
    of features. I'm a simple kind of guy, and don't see the need to allow my
    caller to pick and choose which deprecations they'd like to ignore based
    on some API version.

    Attribute::Deprecated is cute, but only deals with methods, and ironically
    not (what Moose calls) attributes.

    Devel::Deprecation has some pretty nice features, but is more manual than
    I'd like, and again only deals with methods.

    Not to be confused with Moose::Deprecated which can be used to manage
    warnings issued by Moose itself.

    <http://en.wikipedia.org/wiki/Self-deprecation>.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
MooseX-Deprecated
=================

Created:      2013-10-29
Home page:    <https://metacpan.org/release/MooseX-Deprecated>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=MooseX-Deprecated>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.004	2013-10-29

 [ Documentation ]
 - Mention Moose::Deprecated in SEE ALSO.
 - Note Perl 5.8 limitations regarding warnings::warnif().

 [ Test Suite ]
 - Include a cut-down test for Perl 5.8.
 - Skip basic tests in Perl < 5.10; they're not going to work.

0.003	2013-10-29

 [ Bug Fixes ]
 - Stop relying on warnings::fatal_enabled() function which doesn't exist
   on Perl < 5.12.

 [ Other ]
 - Make the undocumented E* functions handle the sprintf stuff within
   themselves.

0.002	2013-10-29

 - Saner technique for skipping internals when issuing warnings (and also
   when detecting the currently enabled warning categories).

0.001	2013-10-29	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "mark attributes and methods as deprecated",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.013, CPAN::Meta::Converter version 2.120921",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "MooseX-Deprecated",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { recommends => { "Dist::Inkt" => 0.001 } },
                        runtime   => { requires => { "MooseX::Role::Parameterized" => "1.00" } },
                        test      => {
                                       requires => {
                                         "Moose"          => "2.0300",
                                         "Test::Fatal"    => 0.007,
                                         "Test::Moose"    => 0,
                                         "Test::More"     => 0.96,
                                         "Test::Warnings" => 0.005,
                                       },
                                     },
                      },
  "provides"       => {
                        "MooseX::Deprecated" => { file => "lib/MooseX/Deprecated.pm", version => 0.004 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => {
                                          web => "http://rt.cpan.org/Dist/Display.html?Queue=MooseX-Deprecated",
                                        },
                        homepage     => "https://metacpan.org/release/MooseX-Deprecated",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          web  => "https://github.com/tobyink/p5-moosex-deprecated",
                                        },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/MooseX-Deprecated/project",
                      },
  "version"        => 0.004,
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

